### DISCUSSION TITLE
Use GDScript over C++ to develop some built-in editor plugins (aka built-in addons)

### DISCUSSION THREAD

**Discussion by Xrayez** [godotengine/godot-proposals#4068] [Editor]
As you may know, developing editor plugins and user interface in general could be a daunting task in Godot's C++. Do you think that the development could be sped up by using GDScript over C++ for core editor plugin development (or plugins in general?)

### Pros

- Faster development cycle.
- Core developers can freely use Godot's editor itself to develop new *built-in* plugins using GDScript. I know there are some plugins that aim to speed up development similarly, like https://github.com/Zylann/godot_scene_code_converter. It means that oftentimes, the prototyping starts from GDScript. So why wasting time porting code to C++ if we could find a way to *embed* GDScript plugins as part of the editor?

### Cons

- The performance may suffer for plugins that require heavy processing, like sprite conversion tools, lightmap baking etc.
- Perhaps there may be issues with debugging built-in GDScript plugins, because they will reside in-memory rather than filesystem.

## Possible implementation

This will require some core modifications:
- Godot must somehow pack GDScript plugins into executable during compilation of the source code. It means that the main repository shall contain plugins written in GDScript, as well as in C++ (for those plugins that require good performance).
- Godot must be able to activate the built-in plugins. Perhaps those plugins could be displayed as virtual and read-only to the filesystem, so that script debugger could be used for those as well.
- Alternatively, plugins could be extracted from memory to the filesystem. But I think this would be intrusive to users.

## Extensibility

This idea could be extended for embedding any kind of plugin for third-party developers:

- goostengine/goost#173

Perhaps this could be extended to GDExtension itself, to allow distributing extensions as part of custom Godot builds (so that there would be no need to download anything extra for each project you're working on).

**Comment by fire** [comment]
Is GDScript embedded addons an idea, a prototype or working in an example?

**Reply by Xrayez** [reply]
I haven't done anything in this direction, so this is mostly conceptual. I do think it's something that can be realistically done with enough dedication.

**Reply by azur-wolve** [reply]
Wouldn't the GDExtension approach of "coupling against the engine as library" (if I've understood correctly) help?

**Reply by Xrayez** [reply]
Both GDScript and GDNative/GDExtension would require downloading them from external source. This means that you'd literally have to copy-paste addons into *every* project you're working on, which is quite a hassle.

This is in contrast to static C++ modules where features can be embedded as part of a single executable. As far as I know, it won't be possible to use dynamic libraries (GDExtension) that could reside alongside Godot executable, so coupling GDExstension with the engine binary wouldn't be possible. The same logic applies to GDScript plugins.

However, this discussion is less about GDExtension but more about GDScript. Due to all of this, the idea is that GDScript plugins could be compiled into Godot executable, for the reasons outlined above. As of now, this would be more useful for engine developers rather than users, because developing official plugins for Godot in GDScript (rather than C++) would certainly speed up the development a lot. This also means that more things could be potentially exposed to GDScript. In some cases, perhaps this could also decrease the binary size, since no C++ templates would need to be used, but just pure GDScript files (which can also be compressed prior to embedding them into Godot executable).

**Reply by azur-wolve** [reply]
> Both GDScript and GDNative/GDExtension would require downloading them from external source. This means that you'd literally have to copy-paste addons into every project you're working on, which is quite a hassle.

What about an addon/module that does it: allowing to "share" between all the projects. Would be such thing possible?

**Reply by Xrayez** [reply]
> What about an addon/module that does it: allowing to "share" between all the projects. Would be such thing possible?

Definitely, look for instance this pull request:

- godotengine/godot#35432

I truly think that support for global plugins should be there (both for editor-only *and* game-specific plugins). But since it got already rejected, I'm proposing alternatives. 😉

**Reply by azur-wolve** [reply]
In specific was meaning an addon (community-made) that allows other addons-modules (whatever) to do that.
But it looks like it needs some look-up on engine-side so..
